### PR TITLE
Move hir_ty to Chalk IR

### LINKED ISSUE #8313: Tracking issue: Move to Chalk IR
We want to get rid of our own representations of types and related data in hir_ty (like `Ty`, `TraitRef` and so on), and instead use the versions from `chalk_ir` everywhere. This is a rather large refactoring, and to achieve it without doing a gigantic PR that takes months, we're gradually aligning our own representations to look like the Chalk versions, with the goal that in the end we can mostly just define `type Ty = chalk_ir::Ty<Interner>` and be done.

We're actually getting close to this goal; I [tried it to see where we are currently](https://github.com/rust-analyzer/rust-analyzer/compare/master...flodiebold:chalk-ir-move?expand=1) and we're at about 400 errors when doing the switch. Going through those errors, I found the following remaining tasks:

- [x] remove interner parameter from `Substitution::interned` (I added it wrongly)
- [x] move `Ty` inherent methods to free-standing, extension trait or builder (except those that exist on Chalk's `Ty` as well)
  - all the `is_xxx` and `as_xxx` and similar accessors should move to extension trait
  - `equals_ctor` and `builtin_deref` can be free-standing in the modules where they're used
  - `substs` and `substs_mut` would ideally be removed
- [x] same for `Substitution` (just `prefix` and `suffix` remaining)
- [x] rename `TyKind::Unknown` -> `Error`
- [x] rename `TyKind::ForeignType` -> `Foreign`
- [x] adjust `Lifetime` to be more like Chalk (i.e. there's an `interned` method that contains the actual enum)
- [x] `self_type_parameter` takes Interner (on `TraitRef` and `ProjectionTy`)
- [x] remove all `Binders` `value` accesses, use `VariableKinds` instead of just a usize
  - make helper function for construction of `VariableKinds` with a number of type vars, since that's all we ever do right now
- [x] align `Binders::subst` with Chalk's `substitute`
- [x] hide `Substitution` internals, which are still getting accessed in lots of places
- [x] add lifetimes/consts to `TyKind::Ref`, `TyKind::Array`, `DynTy`
- [x] `FnPointer`: remove `num_args`, add `num_binders` (those are not the same), add `FnSubst`, `substs` -> `substitution`
- [x] remove `SolutionVariables`
- [x] move `InferenceVar` to `types` module and align it with the Chalk version (`to_inner` / `from_inner` should probably become free-standing functions in `unify.rs`

Note that lots of these can be easily done with [structural search & replace](https://rust-analyzer.github.io/manual.html#structural-search-and-replace)!

### PR DESCRIPTION
**Pull request by flodiebold**
Closes #8313.

There's some further cleanups to do:
 - we're still using our `TypeWalk` in lots of places (not for mutating/folding though, just for walking)
 - we're still using our own canonicalization and unification and our `InferenceTable`
 - ~`ToChalk` still exists and gets called, it's just the identity in most cases now (I'll probably clean those up before merging this)~

### CODE REVIEW

File: crates/hir_ty/src/display.rs [hunk 1]
```diff
@@ -741,7 +745,7 @@ fn write_bounds_like_dyn_trait(
                 if !first {
                     write!(f, " + ")?;
                 }
-                // We assume that the self type is $0 (i.e. the
+                // We assume that the self type is  (i.e. the
```
Comment (flodiebold) [inline crates/hir_ty/src/display.rs:748]: o_O... I think that may have been the result of applying a code action? if there's a $0 in code, do we just let it get treated as a snippet marker?

Comment (matklad) [reply crates/hir_ty/src/display.rs:748]: yup, we don't do escaping. 

:thinking: :imp: I wonder if we can bypass the whole issue by using some unicode private range  character internaly for  `$`, and do escaping in the `to_proto`?

**Comment by flodiebold** [comment]
changelog internal Type inference now uses Chalk's types everywhere, instead of having its own representations and mapping between them

**Review by lnicola — APPROVED** [review]

**Comment by lnicola** [comment]
987 MB -> 1018 MB RAM on self.

**Pull request merged**